### PR TITLE
Fix the execution delay for the jobs (409)

### PR DESCRIPTION
**Pull request by alirezazamani**
**Issues**
Fix the execution delay for the jobs #409 .

**Description**
In the Task Framework part of helix, execution delay for the jobs is not respected.
In this commit, when the job is extracted from the inflighJobs queue, the timeline has been checked before scheduling.

**Tests**
The following tests are written for this issue:
TestExecutionDelay.java

Test Result 1: mvn test
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestAlertingRebalancerFailure.testTagSetIncorrect:174->checkResourceBestPossibleCalFailureState:310 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 843, Failures: 1, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  54:26 min
[INFO] Finished at: 2019-08-21T10:00:54-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-core: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/anajari/my_repos/helix/helix-core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException

Test Result 2: mvn test -Dtest="TestAlertingRebalancerFailure"
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.05 s - in org.apache.helix.integration.TestAlertingRebalancerFailure
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.735 s
[INFO] Finished at: 2019-08-21T10:26:21-07:00
[INFO] ------------------------------------------------------------------------

**Commits**
My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines:
  - Subject is separated from body by a blank line
  - Subject is limited to 50 characters 
  - Subject does not end with a period
  - Subject uses the imperative mood ("add", not "adding")
  - Body wraps at 72 characters
  - Body explains "what" and "why", not "how"

**Code Quality**
My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.helix.task.TaskUtil;
+import org.apache.helix.task.Workflow;
+import org.apache.helix.task.WorkflowContext;
+import org.apache.helix.tools.ClusterVerifiers.BestPossibleExternalViewVerifier;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class TestExecutionDelay extends TaskTestBase {
+  private HelixAdmin admin;
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Let's use an underscore for private variables.
`admin -> _admin`

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  @BeforeClass public void beforeClass() throws Exception {
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Did you format the diff with helix-style? @BeforeClass should take up a line to itself.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    admin = _gSetupTool.getClusterManagementTool();
+    super.beforeClass();
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Minor: usually superclass's methods are called first.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  @Test(expectedExceptions = HelixException.class) public void testExecutionDelay()
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Style - annotation should take up a line on its own

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    builder.addParentChildDependency("JOB0", "JOB1");
+    builder.addParentChildDependency("JOB0", "JOB2");
+    builder.addJob("JOB0", jobBuilder.setExecutionDelay(100000000L));
+    builder.addJob("JOB1", jobBuilder2);
+    builder.addJob("JOB2", jobBuilder3);
+
+    _driver.start(builder.build());
+
+    _driver.pollForWorkflowState(workflowName, 15000, TaskState.COMPLETED);
+
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Could we remove this empty line? :)

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Could we remove this empty line? :)

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    builder.addParentChildDependency("JOB0", "JOB1");
+    builder.addParentChildDependency("JOB0", "JOB2");
+    builder.addJob("JOB0", jobBuilder.setExecutionDelay(100000000L));
+    builder.addJob("JOB1", jobBuilder2);
+    builder.addJob("JOB2", jobBuilder3);
+
+    _driver.start(builder.build());
+
+    _driver.pollForWorkflowState(workflowName, 15000, TaskState.COMPLETED);
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: For longs, please append "L" at the end.
`15000 -> 15000L`

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java (hunk 1 above)
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Where is this HelixAdmin used?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder3 =
+        JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG).setMaxAttemptsPerTask(1)
+            .setWorkflow(workflowName)
+            .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    builder.addParentChildDependency("JOB0", "JOB1");
+    builder.addParentChildDependency("JOB0", "JOB2");
+    builder.addJob("JOB0", jobBuilder.setExecutionDelay(100000000L));
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Magic number -> make it a constant?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java (hunk 7 above)
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Magic number - make it a constant?

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 9]
```diff
@@ -165,7 +165,9 @@ public void updateWorkflowStatus(String workflow, WorkflowConfig workflowCfg,
     RuntimeJobDag runtimeJobDag = _clusterDataCache.getTaskDataCache().getRuntimeJobDag(workflow);
     if (runtimeJobDag != null) {
       for (String inflightJob : runtimeJobDag.getInflightJobList()) {
-        processJob(inflightJob, currentStateOutput, bestPossibleOutput, workflowCtx);
+        if (System.currentTimeMillis() >= workflowCtx.getJobStartTime(inflightJob)) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: For the execution delay itself, adding this check may do the trick. However, two more things to consider:
1. Would this be the right place to place the check? Say, should we include this check inside processJob()? Or should we create a separate method that checks whether a job is eligible to be processed?
2. Could there be more checks we need to perform other than execution delay? It would be great if we could make this a comprehensive fix :)

**Comment by narendly** [comment]
A comment on the test you added. I had a hard time understanding what you're trying to test in your test. I think we could make the test more robust by adding a more intuitive test case, and also add some description in your test so that anyone could see what you're doing clearly.

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java (hunk 9 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: Thank you for your comments.
1- Based on same logic that is added few line below (which calculates the start time and store these information and process the job if possible), this is the place that this check should happen. (same as the scheduleJobs method in WorkflowDistpatcher).
2- About the comprehensive change, I agree with you. This is just a quick fix. We will need to revisit several concepts such as inflight jobs and retry delay. I am working on the retry delay and will create new issue for that as well. After that we can think on a comprehensive changes to the Task Framework.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java (hunk 3 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: I fixed that. Thanks.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java (hunk 4 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Fixed.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java (hunk 1 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Not used. I deleted it.

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 10]
```diff
@@ -602,4 +590,28 @@ private void removeContextsAndPreviousAssignment(String workflow, Set<String> jo
...
+        calculatedStartTime += jobConfig.getExecutionDelay();
+      }
+      calculatedStartTime = Math.max(calculatedStartTime, jobConfig.getExecutionStart());
+      workflowCtx.setJobStartTime(job, calculatedStartTime);
+    }
+    return calculatedStartTime;
+  }
+
+  private boolean isJobReadyToExecute(WorkflowContext workflowCtx, String job) {
+    if (System.currentTimeMillis() < workflowCtx.getJobStartTime(job)) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: Simplify the boolean condition?

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 11]
```diff
@@ -269,29 +271,15 @@ private void scheduleJobs(String workflow, WorkflowConfig workflowCfg,
           clusterDataCache, clusterDataCache.getAssignableInstanceManager())) {
         JobConfig jobConfig = jobConfigMap.get(job);
 
-        // Since the start time is calculated base on the time of completion of parent jobs for this
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: Could you preserve this comment? I think it is valuable.

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 12]
```diff
@@ -602,4 +590,28 @@ private void removeContextsAndPreviousAssignment(String workflow, Set<String> jo
     }
     cache.removeContext(workflow);
   }
+
+  private long calculateStartTime(WorkflowContext workflowCtx, String job, JobConfig jobConfig) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: I think computeStartTimeForJob or something like that is more intuitive and specific. What do you think?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.helix.task.Workflow;
+import org.apache.helix.task.WorkflowContext;
+import org.apache.helix.tools.ClusterVerifiers.BestPossibleExternalViewVerifier;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * Test To check if the Execution Delay is respected.
+ * If Workflow is completed regardless of execution delay, it means this task should be failed.
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Clarify? This line is confusing..

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+  }
+
+  @Test(expectedExceptions = HelixException.class)
+  public void testExecutionDelay() throws InterruptedException {
+    final long timeOutTime = 15000L;
+    final long executionDelay = 100000000L;
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Is this supposed to be some arbitrarily large number? In that case, Long.MAX_VALUE and add a simple inline comment on how it will be used.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    String workflowName = TestHelper.getTestMethodName();
+    Workflow.Builder builder = new Workflow.Builder(workflowName);
+
+    JobConfig.Builder jobBuilder = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Add a block comment on what DAG you're creating. Something like

```
//       1
//     /    \
//   2      3
```

You get the idea? Helps others understand what you're doing.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+public class TestExecutionDelay extends TaskTestBase {
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+  }
+
+  @Test(expectedExceptions = HelixException.class)
+  public void testExecutionDelay() throws InterruptedException {
+    final long timeOutTime = 15000L;
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: One might wonder, why 15 sec?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    builder.addParentChildDependency("JOB0", "JOB2");
+    builder.addJob("JOB0", jobBuilder.setExecutionDelay(executionDelay));
+    builder.addJob("JOB1", jobBuilder2);
+    builder.addJob("JOB2", jobBuilder3);
+
+    // Since the execution delay is set as a long time, it is expected to reach time out.
+    // If the workflow completed without any exception, it means that ExecutionDelay has not been
+    // respected.
+    _driver.start(builder.build());
+    _driver.pollForWorkflowState(workflowName, timeOutTime, TaskState.COMPLETED);
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: I think there's a better way to test this. Instead of relying on pollForWorkflowState timing out, could you check the status of the workflow **and** the individual jobs? What if the controller is very slow and takes 15 seconds to schedule jobs?

One way of not relying on timing is
1. Start workflow
2. Wait until it is in progress and job 0 has been looked at (controller tried to schedule)
3. Read the context and check start time is set with the right amount
4. Check that job 0 and its children jobs have not run by checking their states

This way, your test is robust against timing issues, correct? No matter how long controller takes to start and schedule that first job instead of relying on 15 sec, an arbitrarily long enough number? :)

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 18]
```diff
@@ -602,4 +590,28 @@ private void removeContextsAndPreviousAssignment(String workflow, Set<String> jo
...
+      calculatedStartTime = Math.max(calculatedStartTime, jobConfig.getExecutionStart());
+      workflowCtx.setJobStartTime(job, calculatedStartTime);
+    }
+    return calculatedStartTime;
+  }
+
+  private boolean isJobReadyToExecute(WorkflowContext workflowCtx, String job) {
+    if (System.currentTimeMillis() < workflowCtx.getJobStartTime(job)) {
+      return false;
+    } else {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: else clause is not needed

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 19]
```diff
@@ -165,7 +165,9 @@ public void updateWorkflowStatus(String workflow, WorkflowConfig workflowCfg,
     RuntimeJobDag runtimeJobDag = _clusterDataCache.getTaskDataCache().getRuntimeJobDag(workflow);
     if (runtimeJobDag != null) {
       for (String inflightJob : runtimeJobDag.getInflightJobList()) {
-        processJob(inflightJob, currentStateOutput, bestPossibleOutput, workflowCtx);
+        if (isJobReadyToExecute(workflowCtx, inflightJob)) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: I am OK to fix it right now. But please put a note or TODO here. Adding non-scheduled job to inflight job list make people conceptually confused.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: @dasahcc I discussed with @alirezazamani offline. I feel that we just neglected all the checking required before calling `processJob` on inflight jobs. "Inflight" in this sense doesn't mean that a job is necessarily running; rather, it means that the job has been passed over onto the Controller and now out of RuntimeJobDAG's responsibility. Otherwise we end up doing a lot of "passing jobs back and forth" for every failure/delay situation that will complicate the code/logic even further.

So in short, it was a simple bug that we didn't check before doing processJob here.

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 20]
```diff
@@ -602,4 +590,33 @@ private void removeContextsAndPreviousAssignment(String workflow, Set<String> jo
...
+      calculatedStartTime = Math.max(calculatedStartTime, jobConfig.getExecutionStart());
+      workflowCtx.setJobStartTime(job, calculatedStartTime);
+    }
+    return calculatedStartTime;
+  }
+
+  private boolean isJobReadyToExecute(WorkflowContext workflowCtx, String job) {
+    long calculatedStartTime = workflowCtx.getJobStartTime(job);
+    long timeNow = System.currentTimeMillis();
+    if (timeNow < calculatedStartTime) {
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: This method could be simplified to:

`return System.currentTimeMillis() >= workflowCtx.getJobStartTime(job);`

Also, are we sure that there's no other checks required in this function? If it's just going to be a one-liner (in other words, there are no other checks to perform than checking the calculatedStartTime against current time to determine if a job is ready to be processed/updated), and if the usage is only in this place, perhaps it may not be worth it to make it a separate function. If that's the case, let's just keep what we had inline.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+    builder.addJob("JOB2", jobBuilder3);
+
+    // Since the execution delay is set as a long time, it is expected to reach time out.
+    // If the workflow completed without any exception, it means that ExecutionDelay has not been
+    // respected.
+    _driver.start(builder.build());
+
+    // Verify workflow Context is created.
+    // Wait until Workflow Context is created.
+    boolean resultWorkflow = TestHelper.verify(() -> {
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Minor: 
Boolean naming convention
https://stackoverflow.com/questions/40339847/variable-and-method-naming-conventions-for-boolean-verbs-in-java

workflowCreated might be a better option here?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+    // Wait until startTime is set.
+    boolean resultStartTime = TestHelper.verify(() -> {
+      WorkflowContext wCtx1 = _driver.getWorkflowContext(workflowName);
+      long startTime = -1;
+      try {
+        startTime = wCtx1.getJobStartTime(TaskUtil.getNamespacedJobName(workflowName, "JOB0"));
+      } catch (Exception e) {
+        // pass
+      }
+      return (startTime != -1);
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: parentheses not needed.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    }, 60 * 1000);
+    Assert.assertTrue(resultStartTime);
+
+    // Verify Job Context is not created. If Job Context is created, it means that the execution
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Explain that job context creation implies that the job ran.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+    // delay is not respected. if not created, it means that the execution delay is respected.
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: is respected because the absence of JobContext implies that the job did not run.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+    boolean resultJob = TestHelper.verify(() -> {
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Workflow context at this point exists and the start time has been set. Is there a reason you're using `TestHelper.verify()?` You don't need it here

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+      JobContext jobCtx =
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Verifying the absence of JobContexts for JOB0 makes sense, but what about JOB1 and JOB2?

Alternatively, you should also do a check for the status using getJobState() or getJobStates()

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java (hunk 20 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java]: There is no other checks it needed. I agree. I moved it inline to the caller function.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java (hunk 25 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: @narendly In this case we need this verify. The reason is to avoid using Thread.sleep and then check the jobstate using getJobState. It takes some time for the controller to write the job context and job state to the zookeeper. There are cases where startTime and workflow context is set but the job context is empty and getJobState returns null.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java (hunk 26 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: Fixed. In new commit I am checking if there is no JobContext for all the jobs.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+      JobContext jobCtx1 =
+          _driver.getJobContext(TaskUtil.getNamespacedJobName(workflowName, "JOB1"));
+      JobContext jobCtx2 =
+          _driver.getJobContext(TaskUtil.getNamespacedJobName(workflowName, "JOB2"));
+      return (jobCtx0 != null || jobCtx1 != null || jobCtx2 != null);
+    }, 30 * 1000);
+    Assert.assertFalse(isJobStarted);
+
+
+    Assert.assertNull(_driver.getWorkflowContext(workflowName).getJobState(TaskUtil.getNamespacedJobName(workflowName, "JOB0")));
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestExecutionDelay.java]: We seem to be making a ZK read call for each check. Is it possible to just do one read and do the checks on one read?

**Comment by alirezazamani** [comment]
@narendly mvn test results have been updated in PR.

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @narendly.
